pyro/distributions/extended.py

```diff
@@ -0,0 +1,37 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import math
+
+from pyro.distributions import constraints
+
+from .conjugate import BetaBinomial
+from .torch import Binomial
+
+
+class ExtendedBinomial(Binomial):
+    """
+    EXPERIMENTAL :class:`~pyro.distributions.Binomial` distribution extended to
+    have logical support the entire integers. Numerical support is still the
+    integer interval ``[0, total_count]``
+    """
+    support = constraints.integer
+
+    def log_prob(self, value):
+        result = super().log_prob(value)
+        invalid = ~super().support.check(value)
+        return result.masked_fill(invalid, -math.inf)
+
+
+class ExtendedBetaBinomial(BetaBinomial):
+    """
+    EXPERIMENTAL :class:`~pyro.distributions.BetaBinomial` distribution
+    extended to have logical support the entire integers. Numerical support is
+    still the integer interval ``[0, total_count]``
+    """
+    support = constraints.integer
+
+    def log_prob(self, value):
+        result = super().log_prob(value)
+        invalid = ~super().support.check(value)
```

Comment:
IIUC..without `validate_args`, this should return `-inf` for values outside the support, but will also not do any sample validation (e.g. check that the values are non-integers). So this change lets us use our existing sample validation logic but relaxes it for `log_prob` so that any integer values outside the support are assigned `-inf`?

Reply:
Correct. The difference from NumPyro is that, when `validate_args=True`, there is still some (slightly weaker) validation of shapes and integer values.